npm/podController.go

```diff
@@ -154,154 +163,353 @@ func appendNamedPortIpsets(ipsMgr *ipsm.IpsetManager, portList []v1.ContainerPor
+	// it should exist?
+	cachedPodObj, podExists := c.npMgr.PodMap[key]
+	if !podExists {
+		return
+	}
+
+	// if the podIp exists, it must match the cachedIp
+	if len(podObj.Status.PodIP) > 0 && cachedPodObj.PodIP != podObj.Status.PodIP {
+		metrics.SendErrorLogAndMetric(util.PodID, "[DeletePod] Info: Unexpected state. Pod (Namespace:%s, Name:%s, uid:%s, has cachedPodIp:%s which is different from PodIp:%s",
+			util.GetNSNameWithPrefix(podObj.Namespace), podObj.ObjectMeta.Name, podObj.UID, cachedPodObj.PodIP, podObj.Status.PodIP)
+	}
+
+	c.workqueue.Add(key)
+}
+
+func (c *podController) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	// Start the informer factories to begin populating the informer caches
+	log.Logf("Starting Pod controlle\n")
+
+	log.Logf("Starting workers")
+	// Launch two workers to process Pod resources
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	log.Logf("Started workers")
+	<-stopCh
+	log.Logf("Shutting down workers")
+
+	return nil
+}
+
+func (c *podController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *podController) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer c.workqueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			// As the item in the workqueue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			c.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncHandler, passing it the namespace/name string of the
+		// Pod resource to be synced.
+
+		// TODO : can consider using "c.queue.AddAfter(key, *requeueAfter)" according to error type
+		if err := c.syncPod(key); err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			c.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		log.Logf("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncPod compares the actual state with the desired, and attempts to converge the two.
+func (c *podController) syncPod(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
 		return nil
 	}
 
+	// Get the Pod resource with this namespace/name
+	pod, err := c.podLister.Pods(namespace).Get(name)
```

Comment:
>podLister [](start = 15, length = 9)

Why do we need PodLister?

Reply:
We will get pods from shareInformer's local cache.